File: app/src/androidTest/java/com/bopr/android/smailer/CallProcessorTest.java
package com.bopr.android.smailer;

import android.accounts.AccountsException;
import android.content.Context;
import android.content.SharedPreferences;

import org.junit.Test;

import java.io.IOException;

import static com.bopr.android.smailer.PhoneEvent.REASON_ACCEPTED;
import static com.bopr.android.smailer.PhoneEvent.REASON_TRIGGER_OFF;
import static com.bopr.android.smailer.PhoneEvent.STATE_IGNORED;
import static com.bopr.android.smailer.PhoneEvent.STATE_PENDING;
import static com.bopr.android.smailer.PhoneEvent.STATE_PROCESSED;
import static com.bopr.android.smailer.Settings.DEFAULT_CONTENT;
import static com.bopr.android.smailer.Settings.DEFAULT_TRIGGERS;
import static com.bopr.android.smailer.Settings.PREF_EMAIL_CONTENT;
import static com.bopr.android.smailer.Settings.PREF_EMAIL_TRIGGERS;
import static com.bopr.android.smailer.Settings.PREF_NOTIFY_SEND_SUCCESS;
import static com.bopr.android.smailer.Settings.PREF_RECIPIENTS_ADDRESS;
import static com.bopr.android.smailer.Settings.PREF_SENDER_ACCOUNT;
import static java.lang.System.currentTimeMillis;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;
import static org.mockito.Matchers.any;
import static org.mockito.Matchers.anyBoolean;
import static org.mockito.Matchers.anyInt;
import static org.mockito.Matchers.anySetOf;
import static org.mockito.Matchers.eq;
import static org.mockito.Mockito.anyString;
import static org.mockito.Mockito.doAnswer;
import static org.mockito.Mockito.doNothing;
import static org.mockito.Mockito.doThrow;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

/**
 * {@link CallProcessor} tester.
 *
 * @author Boris Pronin (<a href="mailto:boprsoft.dev@gmail.com">boprsoft.dev@gmail.com</a>)
 */
public class CallProcessorTest extends BaseTest {

    private Database database;
    private Context context;
    private GoogleMail transport;
    private Notifications notifications;
    private SharedPreferences preferences;
    private GeoLocator geoLocator;

    @Override
    @SuppressWarnings("ResourceType")
    public void setUp() throws Exception {
        super.setUp();

        preferences = mock(SharedPreferences.class);
        when(preferences.getString(eq(PREF_SENDER_ACCOUNT), anyString())).thenReturn("sender@mail.com");
        when(preferences.getString(eq(PREF_RECIPIENTS_ADDRESS), anyString())).thenReturn("recipient@mail.com");
        when(preferences.getStringSet(eq(PREF_EMAIL_TRIGGERS), anySetOf(String.class))).thenReturn(DEFAULT_TRIGGERS);
        when(preferences.getStringSet(eq(PREF_EMAIL_CONTENT), anySetOf(String.class))).thenReturn(DEFAULT_CONTENT);

        context = mock(Context.class);
        when(context.getContentResolver()).thenReturn(getContext().getContentResolver());
        when(context.getResources()).thenReturn(getContext().getResources());
        when(context.getSharedPreferences(anyString(), anyInt())).thenReturn(preferences);

        geoLocator = mock(GeoLocator.class);
        when(geoLocator.getLocation()).thenReturn(new GeoCoordinates(60, 30));

        transport = mock(GoogleMail.class);
        notifications = mock(Notifications.class);

        database = new Database(getContext(), "test.sqlite"); /* not a mock context here! */
        database.destroy();
    }

    private PhoneEvent newPhoneEvent(boolean missed) {
        long time = currentTimeMillis();
        return new PhoneEvent("+123", true, time, time + 1000, missed,
                "SMS TEXT", null, null, STATE_PENDING, "device", REASON_ACCEPTED, false);
    }

    /**
     * Tests successful processing - mail sent.
     */
    @Test
    public void testProcessMailSent() throws Exception {
        MethodInvocationsCollector initInvocations = new MethodInvocationsCollector();
        MethodInvocationsCollector sendInvocations = new MethodInvocationsCollector();
        MethodInvocationsCollector showErrorInvocations = new MethodInvocationsCollector();

        doAnswer(initInvocations).when(transport).startSession(anyString(), anyString());
        doAnswer(sendInvocations).when(transport).send(any(MailMessage.class));
        doAnswer(showErrorInvocations).when(notifications).showMailError(anyInt(), anyInt());

        PhoneEvent event = newPhoneEvent(true);

        CallProcessor processor = new CallProcessor(context, transport, notifications, database, geoLocator);
        processor.process(event);

        assertTrue(showErrorInvocations.isEmpty());

        assertEquals(1, initInvocations.count());
        String sender = initInvocations.getArgument(0, 0);
        assertEquals("sender@mail.com", sender);

        assertEquals(1, sendInvocations.count());
        MailMessage message = sendInvocations.getArgument(0, 0);
        assertEquals("[SMailer] Missed call from \"+123\"", message.getSubject());

        PhoneEvent savedEvent = database.getEvents().findFirst();
        assertEquals(event.getAcceptor(), savedEvent.getAcceptor());
        assertEquals(event.getStartTime(), savedEvent.getStartTime());
        assertEquals(event.getPhone(), savedEvent.getPhone());
        assertEquals(STATE_PROCESSED, savedEvent.getState());
        assertEquals(REASON_ACCEPTED, savedEvent.getStateReason());
        assertEquals(new GeoCoordinates(60, 30), event.getLocation());
    }

    /**
     * Tests successful processing - event ignored.
     */
    @Test
    public void testProcessIgnored() throws Exception {
        MethodInvocationsCollector initInvocations = new MethodInvocationsCollector();
        MethodInvocationsCollector sendInvocations = new MethodInvocationsCollector();
        MethodInvocationsCollector showErrorInvocations = new MethodInvocationsCollector();

        doAnswer(initInvocations).when(transport).startSession(anyString(), anyString());
        doAnswer(sendInvocations).when(transport).send(any(MailMessage.class));
        doAnswer(showErrorInvocations).when(notifications).showMailError(anyInt(), anyInt());

        PhoneEvent event = newPhoneEvent(false); /* make it not missed (default filter denies it) */

        CallProcessor processor = new CallProcessor(context, transport, notifications, database, geoLocator);
        processor.process(event);

        assertTrue(showErrorInvocations.isEmpty());
        assertTrue(initInvocations.isEmpty());
        assertTrue(sendInvocations.isEmpty());

        PhoneEvent savedEvent = database.getEvents().findFirst();
        assertEquals(event.getAcceptor(), savedEvent.getAcceptor());
        assertEquals(event.getStartTime(), savedEvent.getStartTime());
        assertEquals(event.getPhone(), savedEvent.getPhone());
        assertEquals(STATE_IGNORED, savedEvent.getState());
        assertEquals(REASON_TRIGGER_OFF, savedEvent.getStateReason());
        assertEquals(new GeoCoordinates(60, 30), event.getLocation());
    }

    /**
     * Test processing when sender is not specified.
     */
    @Test
    public void testProcessNoSender() throws Exception {
        MethodInvocationsCollector initInvocations = new MethodInvocationsCollector();
        MethodInvocationsCollector sendInvocations = new MethodInvocationsCollector();
        MethodInvocationsCollector showErrorInvocations = new MethodInvocationsCollector();

        doAnswer(initInvocations).when(transport).startSession(anyString(), anyString());
        doAnswer(sendInvocations).when(transport).send(any(MailMessage.class));
        doAnswer(showErrorInvocations).when(notifications).showMailError(anyInt(), anyInt());

        when(preferences.getString(eq(PREF_SENDER_ACCOUNT), anyString())).thenReturn(null);

        PhoneEvent event = newPhoneEvent(true);

        CallProcessor processor = new CallProcessor(context, transport, notifications, database, geoLocator);
        processor.process(event);

        assertEquals(1, showErrorInvocations.count());
        int notificationText = showErrorInvocations.getArgument(0, 0);
        assertEquals(R.string.no_account_specified, notificationText);

        assertTrue(initInvocations.isEmpty());
        assertTrue(sendInvocations.isEmpty());

        PhoneEvent savedEvent = database.getEvents().findFirst();
        assertEquals(event.getAcceptor(), savedEvent.getAcceptor());
        assertEquals(event.getStartTime(), savedEvent.getStartTime());
        assertEquals(event.getPhone(), savedEvent.getPhone());
        assertEquals(STATE_PENDING, savedEvent.getState());
        assertEquals(REASON_ACCEPTED, savedEvent.getStateReason());
        assertEquals(new GeoCoordinates(60, 30), event.getLocation());
    }

    /**
     * Test processing when no recipients specified.
     */
    @Test
    public void testProcessNoRecipients() throws Exception {
        MethodInvocationsCollector initInvocations = new MethodInvocationsCollector();
        MethodInvocationsCollector sendInvocations = new MethodInvocationsCollector();
        MethodInvocationsCollector showErrorInvocations = new MethodInvocationsCollector();

        doAnswer(initInvocations).when(transport).startSession(anyString(), anyString());
        doAnswer(sendInvocations).when(transport).send(any(MailMessage.class));
        doAnswer(showErrorInvocations).when(notifications).showMailError(anyInt(), anyInt());

        when(preferences.getString(eq(PREF_RECIPIENTS_ADDRESS), anyString())).thenReturn(null);

        PhoneEvent event = newPhoneEvent(true);

        CallProcessor processor = new CallProcessor(context, transport, notifications, database, geoLocator);
        processor.process(event);

        assertEquals(1, showErrorInvocations.count());
        int notificationText = showErrorInvocations.getArgument(0, 0);
        assertEquals(R.string.no_recipients_specified, notificationText);

        assertTrue(initInvocations.isEmpty());
        assertTrue(sendInvocations.isEmpty());

        PhoneEvent savedEvent = database.getEvents().findFirst();
        assertEquals(event.getAcceptor(), savedEvent.getAcceptor());
        assertEquals(event.getStartTime(), savedEvent.getStartTime());
        assertEquals(event.getPhone(), savedEvent.getPhone());
        assertEquals(STATE_PENDING, savedEvent.getState());
        assertEquals(REASON_ACCEPTED, savedEvent.getStateReason());
        assertEquals(new GeoCoordinates(60, 30), event.getLocation());
    }

    /**
     * Tests processing when mail transport produces init error.
     */
    @Test
    public void testProcessTransportInitFailed() throws Exception {
        MethodInvocationsCollector showErrorInvocations = new MethodInvocationsCollector();
        MethodInvocationsCollector sendInvocations = new MethodInvocationsCollector();

        doAnswer(sendInvocations).when(transport).send(any(MailMessage.class));
        doAnswer(showErrorInvocations).when(notifications).showMailError(anyInt(), anyInt());
        doThrow(new AccountsException("Test error")).when(transport).startSession(anyString(), anyString());

        PhoneEvent event = newPhoneEvent(true);

        CallProcessor processor = new CallProcessor(context, transport, notifications, database, geoLocator);
        processor.process(event);

        assertTrue(sendInvocations.isEmpty());

        assertEquals(1, showErrorInvocations.count());
        int notificationText = showErrorInvocations.getArgument(0, 0);
        assertEquals(R.string.account_not_registered, notificationText);

        PhoneEvent savedEvent = database.getEvents().findFirst();
        assertEquals(event.getAcceptor(), savedEvent.getAcceptor());
        assertEquals(event.getStartTime(), savedEvent.getStartTime());
        assertEquals(event.getPhone(), savedEvent.getPhone());
        assertEquals(STATE_PENDING, savedEvent.getState());
        assertEquals(REASON_ACCEPTED, savedEvent.getStateReason());
        assertEquals(new GeoCoordinates(60, 30), event.getLocation());
    }

    /**
     * Tests processing when mail transport produces send error.
     */
    @Test
    public void testProcessTransportSendFailed() throws Exception {
        MethodInvocationsCollector initInvocations = new MethodInvocationsCollector();
        MethodInvocationsCollector sendInvocations = new MethodInvocationsCollector();
        MethodInvocationsCollector showErrorInvocations = new MethodInvocationsCollector();

        doAnswer(initInvocations).when(transport).startSession(anyString(), anyString());
        doAnswer(sendInvocations).when(transport).send(any(MailMessage.class));
        doAnswer(showErrorInvocations).when(notifications).showMailError(anyInt(), anyInt());
        doThrow(new IOException("Test error")).when(transport).send(any(MailMessage.class));

        PhoneEvent event = newPhoneEvent(true);

        CallProcessor processor = new CallProcessor(context, transport, notifications, database, geoLocator);
        processor.process(event);

        assertEquals(1, initInvocations.count());
        assertTrue(sendInvocations.isEmpty());
        assertTrue(showErrorInvocations.isEmpty());

        PhoneEvent savedEvent = database.getEvents().findFirst();
        assertEquals(event.getAcceptor(), savedEvent.getAcceptor());
        assertEquals(event.getStartTime(), savedEvent.getStartTime());
        assertEquals(event.getPhone(), savedEvent.getPhone());
        assertEquals(STATE_PENDING, savedEvent.getState());
        assertEquals(REASON_ACCEPTED, savedEvent.getStateReason());
        assertEquals(new GeoCoordinates(60, 30), event.getLocation());
    }

    /**
     * When settings goes back to normal last error notification should be removed.
     */
    @Test
    public void testClearNotifications() throws Exception {
        MethodInvocationsCollector showInvocations = new MethodInvocationsCollector();
        MethodInvocationsCollector hideInvocations = new MethodInvocationsCollector();

        doAnswer(showInvocations).when(notifications).showMailError(anyInt(), anyInt());
        doAnswer(hideInvocations).when(notifications).hideAllErrors();

        CallProcessor processor = new CallProcessor(context, transport, notifications, database, geoLocator);

        /* error while sending produces error notification */
        doThrow(new IOException("Test error")).when(transport).send(any(MailMessage.class));
        processor.process(newPhoneEvent(true));

        assertTrue(showInvocations.isEmpty());
        assertTrue(hideInvocations.isEmpty());

        showInvocations.reset();
        hideInvocations.reset();

        /* sending without errors hides all previous error notifications */
        doNothing().when(transport).send(any(MailMessage.class));
        processor.process(newPhoneEvent(true));

        assertTrue(showInvocations.isEmpty());
        assertEquals(1, hideInvocations.count());
    }

    /**
     * When {@link Settings#PREF_NOTIFY_SEND_SUCCESS} setting is set to true success notification should be shown.
     */
    @Test
    public void testSuccessNotification() {
        MethodInvocationsCollector showSuccessInvocation = new MethodInvocationsCollector();

        doAnswer(showSuccessInvocation).when(notifications).showMessage(R.string.email_send, Notifications.ACTION_SHOW_MAIN);

        CallProcessor processor = new CallProcessor(context, transport, notifications, database, geoLocator);

        /* the setting is OFF */
        when(preferences.getBoolean(eq(PREF_NOTIFY_SEND_SUCCESS), anyBoolean())).thenReturn(false);

        processor.process(newPhoneEvent(true));

        assertTrue(showSuccessInvocation.isEmpty());

        /* the setting is ON */
        when(preferences.getBoolean(eq(PREF_NOTIFY_SEND_SUCCESS), anyBoolean())).thenReturn(true);

        processor.process(newPhoneEvent(true));

        assertEquals(1, showSuccessInvocation.count());
        int notificationText = showSuccessInvocation.getArgument(0, 0);
        assertEquals(R.string.email_send, notificationText);
    }

    /**
     * Test resending pending messages.
     */
    @Test
    public void testProcessPending() throws Exception {
        MethodInvocationsCollector showErrorInvocations = new MethodInvocationsCollector();

        doAnswer(showErrorInvocations).when(notifications).showMailError(anyInt(), anyInt());

        /* disable transport */
        doThrow(new IOException("Test error")).when(transport).send(any(MailMessage.class));

        CallProcessor processor = new CallProcessor(context, transport, notifications, database, geoLocator);

        processor.process(newPhoneEvent(true));
        processor.process(newPhoneEvent(true));
        processor.process(newPhoneEvent(true));

        assertEquals(3, database.getEvents().getCount());
        assertEquals(3, database.getPendingEvents().getCount());
        assertTrue(showErrorInvocations.isEmpty());

        /* try resend with disabled transport */

        showErrorInvocations.reset();

        processor.processPending();

        assertEquals(3, database.getEvents().getCount());
        assertEquals(3, database.getPendingEvents().getCount());
        assertTrue(showErrorInvocations.isEmpty()); /* no error notifications should be shown */

        /* enable transport an try again */

        doNothing().when(transport).send(any(MailMessage.class));

        showErrorInvocations.reset();

        processor.processPending();

        assertEquals(3, database.getEvents().getCount());
        assertEquals(0, database.getPendingEvents().getCount());
        assertTrue(showErrorInvocations.isEmpty());
    }

}

File: app/src/androidTest/kotlin/com/bopr/android/smailer/CallProcessorTest.kt
package com.bopr.android.smailer

import android.Manifest.permission.READ_CONTACTS
import android.accounts.Account
import android.accounts.AccountManager
import android.content.Context
import android.content.Context.ACCOUNT_SERVICE
import android.content.Context.CONNECTIVITY_SERVICE
import android.content.SharedPreferences
import android.net.ConnectivityManager
import androidx.test.filters.SmallTest
import androidx.test.rule.GrantPermissionRule
import com.bopr.android.smailer.PhoneEvent.Companion.STATE_IGNORED
import com.bopr.android.smailer.PhoneEvent.Companion.STATE_PENDING
import com.bopr.android.smailer.PhoneEvent.Companion.STATE_PROCESSED
import com.bopr.android.smailer.PhoneEvent.Companion.STATUS_ACCEPTED
import com.bopr.android.smailer.PhoneEvent.Companion.STATUS_TRIGGER_OFF
import com.bopr.android.smailer.Settings.Companion.DEFAULT_EMAIL_CONTENT
import com.bopr.android.smailer.Settings.Companion.DEFAULT_TRIGGERS
import com.bopr.android.smailer.Settings.Companion.PREF_EMAIL_CONTENT
import com.bopr.android.smailer.Settings.Companion.PREF_EMAIL_LOCALE
import com.bopr.android.smailer.Settings.Companion.PREF_EMAIL_TRIGGERS
import com.bopr.android.smailer.Settings.Companion.PREF_NOTIFY_SEND_SUCCESS
import com.bopr.android.smailer.Settings.Companion.PREF_RECIPIENTS_ADDRESS
import com.bopr.android.smailer.Settings.Companion.PREF_SENDER_ACCOUNT
import com.bopr.android.smailer.Settings.Companion.VAL_PREF_DEFAULT
import com.nhaarman.mockitokotlin2.*
import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.mockito.ArgumentMatchers.anyInt
import org.mockito.ArgumentMatchers.anyString
import java.io.IOException
import java.lang.System.currentTimeMillis

/**
 * [CallProcessor] tester.
 *
 * @author Boris Pronin ([boprsoft.dev@gmail.com](mailto:boprsoft.dev@gmail.com))
 */
@SmallTest
class CallProcessorTest : BaseTest() {

    @get:Rule
    var permissionRule: GrantPermissionRule = GrantPermissionRule.grant(READ_CONTACTS)

    private lateinit var database: Database
    private lateinit var context: Context
    private lateinit var transport: GoogleMail
    private lateinit var notifications: Notifications
    private lateinit var preferences: SharedPreferences
    private lateinit var geoLocator: GeoLocator
    private lateinit var processor: CallProcessor
    private lateinit var accountManager: AccountManager

    private fun testingEvent(text: String? = "Message"): PhoneEvent {
        val time = currentTimeMillis()
        return PhoneEvent(
                phone = "+123",
                isIncoming = true,
                startTime = time,
                endTime = time + 1000,
                isMissed = false,
                text = text,
                acceptor = "device"
        )
    }

    @Suppress("DEPRECATION")
    @Before
    fun setUp() {
        preferences = mock {
            on { getString(eq(PREF_SENDER_ACCOUNT), anyOrNull()) }.doReturn("sender@mail.com")
            on { getString(eq(PREF_RECIPIENTS_ADDRESS), anyOrNull()) }.doReturn("recipient@mail.com")
            on { getString(eq(PREF_EMAIL_LOCALE), anyOrNull()) }.doReturn(VAL_PREF_DEFAULT)
            on { getStringSet(eq(PREF_EMAIL_TRIGGERS), anyOrNull()) }.doReturn(DEFAULT_TRIGGERS)
            on { getStringSet(eq(PREF_EMAIL_CONTENT), anyOrNull()) }.doReturn(DEFAULT_EMAIL_CONTENT)
        }

        accountManager = mock {
            on { getAccountsByType(eq("com.google")) }.doReturn(arrayOf(Account("sender@mail.com", "com.google")))
        }

        val connectivityManager = mock<ConnectivityManager> {
            val networkInfo = mock<android.net.NetworkInfo> { /* do not inline. do not optimize imports */
                on { isConnectedOrConnecting }.doReturn(true)
            }
            on { activeNetworkInfo }.doReturn(networkInfo)
        }

        context = mock {
            on { contentResolver }.doReturn(targetContext.contentResolver)
            on { resources }.doReturn(targetContext.resources)
            on { getSharedPreferences(anyString(), anyInt()) }.doReturn(preferences)
            on { getSystemService(eq(ACCOUNT_SERVICE)) }.doReturn(accountManager)
            on { getSystemService(eq(CONNECTIVITY_SERVICE)) }.doReturn(connectivityManager)
        }

        geoLocator = mock {
            on { getLocation() }.doReturn(GeoCoordinates(60.0, 30.0))
        }

        transport = mock()
        notifications = mock()

        Database.databaseName = "test.sqlite"
        targetContext.deleteDatabase(Database.databaseName)
        database = Database(targetContext) /* not a mock context here! */

        processor = CallProcessor(context, database, transport, notifications, geoLocator)
    }

    @After
    fun tearDown() {
        database.close()
    }

    /**
     * Tests successful processing - mail sent.
     */
    @Test
    fun testProcessMailSent() {
        val event = testingEvent()
        processor.process(event)

        verify(transport).send(argThat {
            id == null
                    && subject == "[SMailer] Incoming SMS from \"+123\""
                    && !body.isNullOrBlank()
                    && attachment == null
                    && recipients == "recipient@mail.com"
                    && replyTo == null
                    && from == "sender@mail.com"
        })
        verify(notifications, never()).showSenderAccountError()
        verify(notifications, never()).showRecipientsError(anyInt())
        verify(notifications, never()).showGoogleAccessError()

        val savedEvent = database.events.first()

        assertEquals(event.acceptor, savedEvent.acceptor)
        assertEquals(event.startTime, savedEvent.startTime)
        assertEquals(event.phone, savedEvent.phone)
        assertEquals(STATE_PROCESSED, savedEvent.state)
        assertEquals(STATUS_ACCEPTED, savedEvent.processStatus)
        assertEquals(GeoCoordinates(60.0, 30.0), event.location)
    }

    /**
     * Tests successful processing - event ignored.
     */
    @Test
    fun testProcessIgnored() {
        /* make the event not an SMS then default filter will deny it */
        val event = testingEvent(null)

        processor.process(event)

        verify(transport, never()).send(any())
        verify(notifications, never()).showSenderAccountError()
        verify(notifications, never()).showRecipientsError(anyInt())
        verify(notifications, never()).showGoogleAccessError()

        val savedEvent = database.events.first()

        assertEquals(STATE_IGNORED, savedEvent.state)
        assertEquals(STATUS_TRIGGER_OFF, savedEvent.processStatus)
    }

    /**
     * Test processing when sender account setting is not specified.
     */
    @Test
    fun testProcessNoSender() {
        whenever(preferences.getString(eq(PREF_SENDER_ACCOUNT), anyOrNull())).thenReturn(null)

        val event = testingEvent()
        processor.process(event)

        verify(transport, never()).login(any(), any())
        verify(transport, never()).send(any())
        verify(notifications).showSenderAccountError()

        val savedEvent = database.events.first()

        assertEquals(STATE_PENDING, savedEvent.state)
        assertEquals(STATUS_ACCEPTED, savedEvent.processStatus)
    }

    /**
     * Test processing when recipients setting is not specified.
     */
    @Test
    fun testProcessNoRecipients() {
        whenever(preferences.getString(eq(PREF_RECIPIENTS_ADDRESS), anyOrNull())).thenReturn(null)

        val event = testingEvent()
        processor.process(event)

        verify(transport, never()).send(any())
        verify(notifications).showRecipientsError(eq(R.string.no_recipients_specified))

        val savedEvent = database.events.first()

        assertEquals(STATE_PENDING, savedEvent.state)
        assertEquals(STATUS_ACCEPTED, savedEvent.processStatus)
    }

    /**
     * Transport exception during send does not produce any notifications.
     */
    @Test
    fun testProcessTransportSendFailed() {
        doThrow(IOException("Test error")).whenever(transport).send(any())

        val event = testingEvent()
        processor.process(event)

        verify(transport).login(any(), any())
        verify(transport).send(any())
        verify(notifications, never()).showSenderAccountError()
        verify(notifications, never()).showRecipientsError(anyInt())
        verify(notifications, never()).showGoogleAccessError()

        val savedEvent = database.events.first()

        assertEquals(STATE_PENDING, savedEvent.state)
        assertEquals(STATUS_ACCEPTED, savedEvent.processStatus)
    }

    /**
     * When [PREF_NOTIFY_SEND_SUCCESS] setting is set to true then success notification should be shown.
     */
    @Test
    fun testSuccessNotification() {
        /* the setting is OFF */
        whenever(preferences.getBoolean(eq(PREF_NOTIFY_SEND_SUCCESS), anyOrNull())).thenReturn(false)

        processor.process(testingEvent())

        verify(notifications, never()).showMailSendSuccess()

        /* the setting is ON */
        whenever(preferences.getBoolean(eq(PREF_NOTIFY_SEND_SUCCESS), anyOrNull())).thenReturn(true)

        processor.process(testingEvent())

        verify(notifications).showMailSendSuccess()
    }

    /**
     * Test resending pending messages.
     */
    @Test
    fun testProcessPending() {
        /* disable transport */
        doThrow(IOException("Test error")).whenever(transport).send(any())

        processor.process(testingEvent())
        processor.process(testingEvent())
        processor.process(testingEvent())

        assertEquals(3, database.events.size)
        assertEquals(3, database.events.filterPending.size)
        verify(notifications, never()).showSenderAccountError()
        verify(notifications, never()).showRecipientsError(anyInt())
        verify(notifications, never()).showGoogleAccessError()

        /* try resend with disabled transport */
        processor.processPending()

        assertEquals(3, database.events.size)
        assertEquals(3, database.events.filterPending.size)
        verify(notifications, never()).showSenderAccountError()
        verify(notifications, never()).showRecipientsError(anyInt())
        verify(notifications, never()).showGoogleAccessError()

        /* enable transport an try again */
        doNothing().whenever(transport).send(any())

        processor.processPending()

        assertEquals(3, database.events.size)
        assertEquals(0, database.events.filterPending.size)
        verify(notifications, never()).showSenderAccountError()
        verify(notifications, never()).showRecipientsError(anyInt())
        verify(notifications, never()).showGoogleAccessError()
    }
}